src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java

```diff
@@ -0,0 +1,452 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks that lambda usage can be replaced with method reference.
+ * </p>
+ * <p>
+ * Rationale: method references are easier to read.
```

Comment:
this is against https://github.com/vavr-io/vavr/issues/2600#issuecomment-631518540 , so Check is controversial.
We need more feedback on this to tune rationale.

Reply:
That is an interesting discussion. However I'd tend to align with both Josh Bloch's Effective Java and SonarSource.

In Effective Java 3rd Edition- Item 43: Prefer method references to lambdas (I don't really know of a good resource to link to for this but I'm staring at the physical book right now)

SonarSource: [Rule 1612](https://rules.sonarsource.com/java/type/Code%20Smell/RSPEC-1612)

Reply:
@mttjj thanks for link.
Link above says that "Method/constructor references are commonly agreed to be, most of the time, more compact and readable than using lambdas, and are therefore preferred." which do not contradict feedback from vavr. since they were saying about some specific cases.
@romani This can be placed as rationale for check. What do you think?

Reply:
Here's the tl;dr of Item 43 of "Effective Java" (emphasis original):
"In summary, method references often provide a more succinct alternative to lambdas. **Where method references are shorter and clearer, use them; where they aren't, stick with lambdas.**"

Reply:
@romani I changed rationale

Reply:
yes, rationale should less strict, as even Sonar has `In some rare cases, when it is not clear from the context what kind of function is being described and reference would not increase the clarity, it might be fine to keep the lambda.`
I agree with vavr that human is not good to read long lines for main part at the end.
I am not sure how to automatically detect this, but life and practice will let us know, in future we might update Check to be more smart.